pkg/network/vlan/vlan.go

```diff
@@ -2,190 +2,130 @@ package vlan
-			return err
-		}
-		if err := v.bridge.ToLink().AddRoutes(l); err != nil {
-			return err
-		}
+		return fmt.Errorf("ensure bridge %s failed, error: %w", v.bridge.Name, err)
 	}
 
 	// set master
 	if err := l.SetMaster(v.bridge, v.vids); err != nil {
 		return err
 	}
-	// delete routes of NIC
-	if err := l.DeleteRoutes(); err != nil {
-		return err
-	}
 
-	v.nic = l
-	v.startMonitor()
+	v.uplink = l
 
-	klog.Info("setup VLAN network successfully")
 	return nil
 }
 
 // Note: It's required to call function GetVlanWithNic before tearing down VLAN.
 func (v *Vlan) Teardown() error {
 	klog.Info("start to tear down VLAN network")
-	if v.nic == nil {
-		return fmt.Errorf("vlan network hasn't attached a NIC")
+	if v.uplink == nil {
+		return fmt.Errorf("bridge %s hasn't attached an uplink", v.bridge.Name)
 	}
 
-	v.stopMonitor()
-
-	// delete iptables rule
-	if err := v.bridge.ToLink().DeleteIptForward(); err != nil {
-		return err
-	}
 	// set no master, VIDs will be auto-removed
-	if err := v.nic.SetNoMaster(); err != nil {
-		return fmt.Errorf("set %s no master failed, error: %w", v.nic.Name(), err)
+	if err := v.uplink.SetNoMaster(); err != nil {
+		return fmt.Errorf("set %s no master failed, error: %w", v.uplink.Attrs().Name, err)
 	}
-	// append route for NIC
-	if err := v.nic.AddRoutes(v.bridge); err != nil {
-		return err
-	}
-	// delete route of bridge explicitly
-	if err := v.bridge.ToLink().DeleteRoutes(); err != nil {
-		return err
+
+	if err := v.uplink.Remove(); err != nil {
+		return fmt.Errorf("delete uplink %s failed, error: %w", v.uplink.Attrs().Name, err)
 	}
-	// delete IPv4 address of bridge
-	if err := v.bridge.ClearAddr(); err != nil {
-		return err
+
+	if err := iface.NewLink(v.bridge).Remove(); err != nil {
+		return fmt.Errorf("delete bridge %s failed, error: %w", v.bridge.Name, err)
 	}
 
 	klog.Info("tear down VLAN network successfully")
 	return nil
 }
 
-func (v *Vlan) startMonitor() {
-	bridgeMonitorHandler := monitor.Handler{
-		NewLink: v.afterLinkDown,
+func (v *Vlan) AddLocalArea(vid uint16, cidr string) error {
+	if v.uplink == nil {
+		return fmt.Errorf("bridge %s hasn't attached an uplink", v.bridge.Name)
 	}
-
-	nicMonitorHandler := monitor.Handler{
-		NewLink: v.afterLinkDown,
-	}
-
-	w := network.GetWatcher()
-
-	w.EmptyLink()
-	w.AddLink(v.bridge.Index(), bridgeMonitorHandler)
-	w.AddLink(v.nic.Index(), nicMonitorHandler)
-}
-
-func (v *Vlan) stopMonitor() {
-	network.GetWatcher().DelLink(v.bridge.Index())
-	network.GetWatcher().DelLink(v.nic.Index())
-}
-
-func (v *Vlan) AddLocalArea(id int, cidr string) error {
-	if v.nic == nil {
-		return fmt.Errorf("physical nic vlan network")
+	if ok, _ := v.findVid(vid); ok {
+		return nil
 	}
 
-	if err := v.nic.AddBridgeVlan(uint16(id)); err != nil {
-		return fmt.Errorf("add bridge vlan %d failed, error: %w", id, err)
+	if err := v.uplink.AddBridgeVlan(vid); err != nil {
+		return fmt.Errorf("add bridge vlanconfig %d failed, error: %w", vid, err)
```

Comment:
In such a path, considering reconciler
```
v.findVid (return when existing)

v.uplink.AddBridgeVlan
v.vids = append(v.vids, vid)
iface.EnsureRouteViaGateway(cidr)
```

1. `v.vids = append(v.vids, vid)` seems should be put in last,  before return nil finally

2. what's the return result of `v.uplink.AddBridgeVlan` for duplicated vlan ?  in case it returns error, then add an vlan id existing check in error case (inside `AddBridgeVlan`)  is roubust.